8.8-rc-1: Daemon Toolchain Support on Windows: WindowsRegistry is not supported on this operating system.

### Current Behavior

In IDEA:
![image](https://github.com/gradle/gradle/assets/22521688/a89c11d3-6f28-4a8b-8372-a5b6144e99c9)

Console: 
```stacktrace
org.gradle.tooling.GradleConnectionException: Could not run phased build action using connection to Gradle distribution 'https://services.gradle.org/distributions/gradle-8.8-rc-1-bin.zip'.
	at org.gradle.tooling.internal.consumer.ExceptionTransformer.transform(ExceptionTransformer.java:55)
	at org.gradle.tooling.internal.consumer.ExceptionTransformer.transform(ExceptionTransformer.java:29)
	at org.gradle.tooling.internal.consumer.ResultHandlerAdapter.onFailure(ResultHandlerAdapter.java:43)
	at org.gradle.tooling.internal.consumer.async.DefaultAsyncConsumerActionExecutor$1$1.run(DefaultAsyncConsumerActionExecutor.java:69)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: org.gradle.internal.nativeintegration.NativeIntegrationUnavailableException: WindowsRegistry is not supported on this operating system.
	at org.gradle.internal.nativeintegration.services.NativeServices$BrokenService.invoke(NativeServices.java:543)
	at jdk.proxy36/jdk.proxy36.$Proxy371.getSubkeys(Unknown Source)
	at org.gradle.jvm.toolchain.internal.WindowsInstallationSupplier.getVersions(WindowsInstallationSupplier.java:79)
	at org.gradle.jvm.toolchain.internal.WindowsInstallationSupplier.find(WindowsInstallationSupplier.java:70)
	at org.gradle.jvm.toolchain.internal.WindowsInstallationSupplier.lambda$findOpenJDKs$2(WindowsInstallationSupplier.java:91)
	at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:273)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:992)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.forEachRemaining(StreamSpliterators.java:310)
	at java.base/java.util.stream.Streams$ConcatSpliterator.forEachRemaining(Streams.java:734)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.gradle.jvm.toolchain.internal.WindowsInstallationSupplier.findInstallationsInRegistry(WindowsInstallationSupplier.java:65)
	at org.gradle.jvm.toolchain.internal.WindowsInstallationSupplier.get(WindowsInstallationSupplier.java:49)
	at org.gradle.jvm.toolchain.internal.WindowsInstallationSupplier.get(WindowsInstallationSupplier.java:31)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$15$1.accept(ReferencePipeline.java:541)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:960)
	at java.base/java.util.stream.ReduceOps$ReduceTask.doLeaf(ReduceOps.java:934)
	at java.base/java.util.stream.AbstractTask.compute(AbstractTask.java:327)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:754)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

### Expected Behavior

Working build

### Context (optional)

Workaround:
Delete the `gradle-daemon-jvm.properties` file

### Steps to Reproduce

0. Use Gradle 8.8-rc-1
1. Run `./gradlew :updateDaemonJvm` on Linux/macOS
2. Checkout the repo on Windows

You can use https://github.com/hfhbd/adventOfCode as a reproducer, but the actual structure/project does not really matter, as long as you use Gradle 8.8-rc-1 with a `gradle-daemon-jvm.properties` file

### Gradle version

8.8-rc-1

### Build scan URL (optional)

_No response_

### Your Environment (optional)

Windows 10
Java Home: eclipse_adoptium-21-amd64-windows\jdk-21.0.3+9

IntelliJ IDEA 2024.1 (Ultimate Edition)
Build #IU-241.14494.240, built on March 28, 2024
Runtime version: 17.0.10+8-b1207.12 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 10.0
GC: G1 Young Generation, G1 Old Generation
Memory: 4096M
Cores: 8
Kotlin: 241.14494.240-IJ

--- a/platforms/core-runtime/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java
+++ b/platforms/core-runtime/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java
@@ -405,10 +405,10 @@ public class NativeServices extends DefaultServiceRegistry implements ServiceReg
         if (useNativeIntegrations && operatingSystem.isWindows()) {
             return net.rubygrapefruit.platform.Native.get(WindowsRegistry.class);
         }
-        return notAvailable(WindowsRegistry.class);
+        return notAvailable(WindowsRegistry.class, operatingSystem);
     }
 
-    public SystemInfo createSystemInfo() {
+    public SystemInfo createSystemInfo(OperatingSystem operatingSystem) {
         if (useNativeIntegrations) {
             try {
                 return net.rubygrapefruit.platform.Native.get(SystemInfo.class);
@@ -416,10 +416,10 @@ public class NativeServices extends DefaultServiceRegistry implements ServiceReg
                 LOGGER.debug("Native-platform system info is not available. Continuing with fallback.");
             }
         }
-        return notAvailable(SystemInfo.class);
+        return notAvailable(SystemInfo.class, operatingSystem);
     }
 
-    protected Memory createMemory() {
+    protected Memory createMemory(OperatingSystem operatingSystem) {
         if (useNativeIntegrations) {
             try {
                 return net.rubygrapefruit.platform.Native.get(Memory.class);
@@ -427,7 +427,7 @@ public class NativeServices extends DefaultServiceRegistry implements ServiceReg
                 LOGGER.debug("Native-platform memory integration is not available. Continuing with fallback.");
             }
         }
-        return notAvailable(Memory.class);
+        return notAvailable(Memory.class, operatingSystem);
     }
 
     protected ProcessLauncher createProcessLauncher() {
@@ -441,7 +441,7 @@ public class NativeServices extends DefaultServiceRegistry implements ServiceReg
         return new DefaultProcessLauncher();
     }
 
-    protected PosixFiles createPosixFiles() {
+    protected PosixFiles createPosixFiles(OperatingSystem operatingSystem) {
         if (useNativeIntegrations) {
             try {
                 return net.rubygrapefruit.platform.Native.get(PosixFiles.class);
@@ -449,7 +449,7 @@ public class NativeServices extends DefaultServiceRegistry implements ServiceReg
                 LOGGER.debug("Native-platform posix files integration is not available. Continuing with fallback.");
             }
         }
-        return notAvailable(UnavailablePosixFiles.class);
+        return notAvailable(UnavailablePosixFiles.class, operatingSystem);
     }
 
     protected HostnameLookup createHostnameLookup() {
@@ -505,7 +505,7 @@ public class NativeServices extends DefaultServiceRegistry implements ServiceReg
         };
     }
 
-    protected FileSystems createFileSystems() {
+    protected FileSystems createFileSystems(OperatingSystem operatingSystem) {
         if (useNativeIntegrations) {
             try {
                 return net.rubygrapefruit.platform.Native.get(FileSystems.class);
@@ -513,11 +513,11 @@ public class NativeServices extends DefaultServiceRegistry implements ServiceReg
                 LOGGER.debug("Native-platform file systems information is not available. Continuing with fallback.");
             }
         }
-        return notAvailable(FileSystems.class);
+        return notAvailable(FileSystems.class, operatingSystem);
     }
 
-    private <T> T notAvailable(Class<T> type) {
-        return Cast.uncheckedNonnullCast(Proxy.newProxyInstance(type.getClassLoader(), new Class<?>[]{type}, new BrokenService(type.getSimpleName())));
+    private <T> T notAvailable(Class<T> type, OperatingSystem operatingSystem) {
+        return Cast.uncheckedNonnullCast(Proxy.newProxyInstance(type.getClassLoader(), new Class<?>[]{type}, new BrokenService(type.getSimpleName(), useNativeIntegrations, operatingSystem)));
     }
 
     private static String format(Throwable throwable) {
@@ -533,14 +533,18 @@ public class NativeServices extends DefaultServiceRegistry implements ServiceReg
 
     private static class BrokenService implements InvocationHandler {
         private final String type;
+        private final boolean useNativeIntegrations;
+        private final OperatingSystem operatingSystem;
 
-        private BrokenService(String type) {
+        private BrokenService(String type, boolean useNativeIntegrations, OperatingSystem operatingSystem) {
             this.type = type;
+            this.useNativeIntegrations = useNativeIntegrations;
+            this.operatingSystem = operatingSystem;
         }
 
         @Override
         public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
-            throw new org.gradle.internal.nativeintegration.NativeIntegrationUnavailableException(String.format("%s is not supported on this operating system.", type));
+            throw new org.gradle.internal.nativeintegration.NativeIntegrationUnavailableException(String.format("Service '%s' is not available (os=%s, enabled=%s).", type, operatingSystem, useNativeIntegrations));
         }
     }
 

--- a/platforms/core-runtime/native/src/test/groovy/org/gradle/internal/nativeintegration/services/NativeServicesTest.groovy
+++ b/platforms/core-runtime/native/src/test/groovy/org/gradle/internal/nativeintegration/services/NativeServicesTest.groovy
@@ -21,11 +21,14 @@ import net.rubygrapefruit.platform.WindowsRegistry
 import org.gradle.api.internal.file.temp.TemporaryFileProvider
 import org.gradle.internal.file.Chmod
 import org.gradle.internal.file.Stat
+import org.gradle.internal.nativeintegration.NativeIntegrationUnavailableException
 import org.gradle.internal.nativeintegration.ProcessEnvironment
 import org.gradle.internal.nativeintegration.console.ConsoleDetector
 import org.gradle.internal.nativeintegration.filesystem.FileCanonicalizer
 import org.gradle.internal.nativeintegration.filesystem.FileSystem
 import org.gradle.internal.os.OperatingSystem
+import org.gradle.test.precondition.Requires
+import org.gradle.test.preconditions.UnitTestPreconditions
 import org.gradle.util.UsesNativeServices
 import spock.lang.Specification
 
@@ -82,5 +85,15 @@ class NativeServicesTest extends Specification {
     def "makes a TemporaryFileProvider available"() {
         expect:
         services.get(TemporaryFileProvider) != null
+    }
+
+    @Requires(UnitTestPreconditions.NotWindows)
+    def "try using a WindowsRegistry on a non-Windows OS"() {
+        def service = services.get(WindowsRegistry)
+        when:
+        service.getSubkeys(WindowsRegistry.Key.HKEY_LOCAL_MACHINE, "SOFTWARE\\AdoptOpenJDK\\JDK")
+        then:
+        def e = thrown(NativeIntegrationUnavailableException)
+        e.message == "Service 'WindowsRegistry' is not available (os=${OperatingSystem.current()}, enabled=true)."
     }
 }

--- a/platforms/jvm/jvm-services/build.gradle.kts
+++ b/platforms/jvm/jvm-services/build.gradle.kts
@@ -43,6 +43,7 @@ dependencies {
     api(libs.nativePlatform)
 
     implementation(project(":functional"))
+    implementation(project(":native"))
 
     implementation(libs.guava)
     implementation(libs.asm)

--- a/platforms/jvm/jvm-services/src/main/java/org/gradle/jvm/toolchain/internal/WindowsInstallationSupplier.java
+++ b/platforms/jvm/jvm-services/src/main/java/org/gradle/jvm/toolchain/internal/WindowsInstallationSupplier.java
@@ -19,6 +19,7 @@ package org.gradle.jvm.toolchain.internal;
 import com.google.common.collect.Lists;
 import net.rubygrapefruit.platform.MissingRegistryEntryException;
 import net.rubygrapefruit.platform.WindowsRegistry;
+import org.gradle.internal.nativeintegration.NativeIntegrationUnavailableException;
 import org.gradle.internal.os.OperatingSystem;
 
 import java.io.File;
@@ -67,9 +68,9 @@ public class WindowsInstallationSupplier implements InstallationSupplier {
 
     private List<String> find(String sdkSubkey, String path, String value) {
         try {
-            return getVersions(sdkSubkey).stream()
-                .map(version -> getValue(sdkSubkey, path, value, version)).collect(Collectors.toList());
-        } catch (MissingRegistryEntryException e) {
+            List<String> versions = getVersions(sdkSubkey);
+            return versions.stream().map(version -> getValue(sdkSubkey, path, value, version)).collect(Collectors.toList());
+        } catch (MissingRegistryEntryException | NativeIntegrationUnavailableException e) {
             // Ignore
             return Collections.emptyList();
         }

--- a/platforms/jvm/jvm-services/src/test/groovy/org/gradle/jvm/toolchain/internal/WindowsInstallationSupplierTest.groovy
+++ b/platforms/jvm/jvm-services/src/test/groovy/org/gradle/jvm/toolchain/internal/WindowsInstallationSupplierTest.groovy
@@ -18,8 +18,7 @@ package org.gradle.jvm.toolchain.internal
 
 import net.rubygrapefruit.platform.MissingRegistryEntryException
 import net.rubygrapefruit.platform.WindowsRegistry
-import org.gradle.api.internal.provider.Providers
-import org.gradle.api.provider.ProviderFactory
+import org.gradle.internal.nativeintegration.NativeIntegrationUnavailableException
 import org.gradle.internal.os.OperatingSystem
 import spock.lang.Specification
 
@@ -74,8 +73,20 @@ class WindowsInstallationSupplierTest extends Specification {
 
     def "handles absent adoptopenjdk keys"() {
         given:
-        registry.getSubkeys(WindowsRegistry.Key.HKEY_LOCAL_MACHINE, "SOFTWARE\\AdoptOpenJDK\\JDK") >> { throw new MissingRegistryEntryException() }
-        def supplier = createSupplier(OperatingSystem.MAC_OS)
+        registry.getSubkeys(WindowsRegistry.Key.HKEY_LOCAL_MACHINE, _) >> { throw new MissingRegistryEntryException() }
+        def supplier = createSupplier()
+
+        when:
+        def locations = supplier.get()
+
+        then:
+        locations.isEmpty()
+    }
+
+    def "gracefully handles unavailable native integration"() {
+        given:
+        registry.getSubkeys(WindowsRegistry.Key.HKEY_LOCAL_MACHINE, _) >> { throw new NativeIntegrationUnavailableException() }
+        def supplier = createSupplier()
 
         when:
         def locations = supplier.get()
@@ -114,11 +125,4 @@ class WindowsInstallationSupplierTest extends Specification {
     WindowsInstallationSupplier createSupplier(OperatingSystem os = OperatingSystem.WINDOWS) {
         new WindowsInstallationSupplier(registry, os)
     }
-
-    ProviderFactory createProviderFactory(String propertyValue) {
-        def providerFactory = Mock(ProviderFactory)
-        providerFactory.gradleProperty("org.gradle.java.installations.auto-detect") >> Providers.notDefined()
-        providerFactory
-    }
-
 }

